File: Lab2/CSharp/QuickSortTests.cs
using System;
using System.Linq;
using FluentAssertions;
using NUnit.Framework;

namespace CSharp {
    [TestFixture]
    public class QuickSortTests {
        private static Func<int, bool> _splitByPreficate = TestSources.SplitByPredicate;
        private static (int[] source, int[] left, int[] right)[] _splitBySource = TestSources.SplitBySource;
        private static int[][] _sortSorce = TestSources.SortSorce;

        [TestCaseSource(nameof(_splitBySource))]
        public void SplitBy_v1LinkedList_Test((int[] source, int[] left, int[] right) args) {
            var result = args.source.SplitBy_v1LinkedList(_splitByPreficate);

            result.left.ToArray().Should().BeEquivalentTo(args.left);
            result.right.ToArray().Should().BeEquivalentTo(args.right);
        }

        [TestCaseSource(nameof(_splitBySource))]
        public void SplitBy_v2List_Test((int[] source, int[] left, int[] right) args) {
            var result = args.source.SplitBy_v2List(_splitByPreficate);

            result.left.ToArray().Should().BeEquivalentTo(args.left);
            result.right.ToArray().Should().BeEquivalentTo(args.right);
        }

        [TestCaseSource(nameof(_sortSorce))]
        public void QuickSortListTest(int[] array) {
            var result = array.QuickSort(Ext.SplitBy_v2List);
            
            var expected = array.OrderBy(x => x).ToArray();
            result.Should().BeEquivalentTo(expected);
        }
        
        [TestCaseSource(nameof(_sortSorce))]
        public void QuickSorLinkedListtTest(int[] array) {
            var result = array.QuickSort(Ext.SplitBy_v1LinkedList);
            
            var expected = array.OrderBy(x => x).ToArray();
            result.Should().BeEquivalentTo(expected);
        }
    }
}

File: Lab2/FSharp/QuickSortTests.fs
module QuickSortTests

open NUnit.Framework
open FsUnit

[<TestFixture>]
type Tests() =
    
    static member SplitBySource = TestSources.splitBySource
    
    static member SplitByPredicate = TestSources.splitByPredicate
    
    static member SortSource = TestSources.sortSource
    
    [<TestCaseSource("SortSource")>]
    static member QuickSortTest list =
        let result = QuickSort.quickSort list

        let expected = List.sort list
        result |> should equal expected

    [<TestCaseSource("SplitBySource")>]
    static member SplitByTest (tuple : 'a list * 'a list * 'a list) =
        let (source, expectedLeft, expectedRight) = tuple 
        let (left, right) = QuickSort.splitBy Tests.SplitByPredicate source
    
        left |> should  equal expectedLeft 
        right |> should equal expectedRight

